src/pseudopeople/entity_types.py

```diff
@@ -82,49 +83,48 @@ def probability_key(self) -> str:
-    ) -> Tuple[pd.Series, pd.Index]:
-        # Do not noise if the column is empty
-        if (data[column_name].notna() & (data[column_name] != "")).sum() == 0:
-            return data[column_name], pd.Index([])
+    ) -> None:
+        if dataset_data.is_empty(column_name):
+            return
+
         noise_level = configuration[
             Keys.CELL_PROBABILITY
-        ] * self.noise_level_scaling_function(data, column_name)
-        # Certain columns have their noise level scaled so we must check to make sure the noise level is within the
-        # allowed range between 0 and 1 for probabilities
+        ] * self.noise_level_scaling_function(dataset_data.data, column_name)
+
+        # Certain columns have their noise level scaled so we must check to make
+        # sure the noise level is within the allowed range between 0 and 1 for
+        # probabilities
         noise_level = min(noise_level, 1.0)
         to_noise_idx = get_index_to_noise(
-            data,
+            dataset_data,
             noise_level,
-            randomness_stream,
             f"{self.name}_{column_name}",
-            is_column_noise=True,
-            missingness=missingness,
+            [column_name] + self.additional_column_getter(column_name),
         )
         if to_noise_idx.empty:
             logger.debug(
-                f"No cells chosen to noise for noise function {self.name} on column {column_name}. "
-                "This is likely due to a combination of the configuration noise levels and the simulated population data."
+                f"No cells chosen to noise for noise function {self.name} on "
+                f"column {column_name}. "
+                "This is likely due to a combination "
+                "of the configuration noise levels and the simulated population "
+                "data."
             )
-            return data[column_name], to_noise_idx
-        noised_data = self.noise_function(
-            data.loc[to_noise_idx],
+            return
+
+        original_dtype_name = dataset_data.data[column_name].dtype.name
+        self.noise_function(
+            dataset_data,
             configuration,
-            randomness_stream,
-            dataset_name,
+            to_noise_idx,
             column_name,
         )
 
+        # todo investigate this and also move the logic inside DatasetData
         # Coerce noised column dtype back to original column's if it has changed
-        if noised_data.dtype.name != data[column_name].dtype.name:
-            noised_data = noised_data.astype(data[column_name].dtype)
-
-        result = data[column_name].copy()
-        result.loc[to_noise_idx] = noised_data
-
-        return result, to_noise_idx
+        if dataset_data.data[column_name].dtype.name != original_dtype_name:
+            dataset_data = dataset_data.data[column_name].astype(
+                dataset_data.data[column_name][column_name].dtype
+            )
```

Comment:
ALl the updates we have made allow us to update dataset_data in place. So noise functions return None. Entity types return none because entity types only serve to pass args to a specific noise function.